client/go/dogma/content_service.go

```diff
@@ -0,0 +1,320 @@
+	}
+	return entries, res, nil
+}
+
+func (con *contentService) getHistory(ctx context.Context,
+	projectName, repoName, from, to, pathPattern string) ([]*Commit, *http.Response, error) {
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/commits/%v", defaultPathPrefix, projectName, repoName, from)
+
+	urlQuery := ""
+	if len(pathPattern) != 0 {
+		urlQuery = fmt.Sprintf("?path=%v", pathPattern)
+		if len(to) != 0 {
+			urlQuery += fmt.Sprintf("&to=%v", to)
+		}
+	} else if len(to) != 0 {
+		urlQuery = fmt.Sprintf("?to=%v", to)
+	}
+	u += urlQuery
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var commits []*Commit
+	res, err := con.client.do(ctx, req, &commits)
+	if err != nil {
+		return nil, res, err
+	}
+	return commits, res, nil
+}
+
+func (con *contentService) getDiff(ctx context.Context,
+	projectName, repoName, from, to string, query *Query) (*Change, *http.Response, error) {
+	path := query.Path
+	if len(path) == 0 {
+		return nil, nil, fmt.Errorf("the path should not be empty")
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/compare", defaultPathPrefix, projectName, repoName)
+	urlQuery, err := getDiffURLQuery(from, to, path, query.Expressions)
+	if err != nil {
+		return nil, nil, err
+	}
+	u += urlQuery
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	change := new(Change)
+	res, err := con.client.do(ctx, req, change)
+	if err != nil {
+		return nil, res, err
+	}
+
+	return change, res, nil
+}
+
+// getDiffURLQuery currently only supports JSON path.
+func getDiffURLQuery(from, to, path string, jsonPaths []string) (urlQuery string, err error) {
+	if len(jsonPaths) != 0 {
+		if urlQuery, err = getJSONPaths(path, jsonPaths); err != nil {
+			return "", err
+		}
+	}
+
+	if !strings.HasPrefix(urlQuery, "?") {
+		urlQuery = fmt.Sprintf("?path=%v", path)
+	} else {
+		urlQuery += fmt.Sprintf("&path=%v", path)
+	}
+
+	if len(from) != 0 {
+		urlQuery += fmt.Sprintf("&from=%v", from)
+	}
+
+	if len(to) != 0 {
+		urlQuery += fmt.Sprintf("&to=%v", to)
+	}
+
+	return urlQuery, nil
+}
+
+func (con *contentService) getDiffs(ctx context.Context,
+	projectName, repoName, from, to, pathPattern string) ([]*Change, *http.Response, error) {
+	if len(pathPattern) == 0 {
+		pathPattern = "/**"
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/compare", defaultPathPrefix, projectName, repoName)
+	urlQuery, err := getDiffURLQuery(from, to, pathPattern, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	u += urlQuery
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var changes []*Change
+	res, err := con.client.do(ctx, req, &changes)
+	if err != nil {
+		return nil, res, err
+	}
+	return changes, res, nil
+}
+
+type Push struct {
+	CommitMessage *CommitMessage `json:"commitMessage"`
+	Changes       []Change       `json:"changes"`
```

Comment:
how about use []*Change instead for uniformity

Reply:
That's a good idea. 
Is there any convenient way to convert the `[]Change` to `[]*Change`?

Reply:
I'm afraird not :(

Reply:
I just changed the method parameter to take `[]*Change`.
I realized that I don't have to take the argument as varargs because I need at least one `Change`.
Thank you~!